packages/core/src/browser/tree/tree-widget.ts

```diff
@@ -229,6 +375,18 @@ export class TreeWidget extends VirtualWidget implements StatefulWidget {
         return Object.assign({}, props, { visible, indentSize });
     }
 
+    protected getDecorations(node: ITreeNode): DecorationData[] {
+        const decorations = this.decorations.get(node.id);
+        if (decorations) {
+            return decorations.sort(DecorationData.compare);
+        }
+        return [];
+    }
+
+    protected getDecorationData<K extends keyof DecorationData>(node: ITreeNode, key: K): DecorationData[K][] {
+        return this.getDecorations(node).filter(data => data[key] !== undefined).map(data => data[key]).filter(notEmpty);
```

Comment:
@akosyakov, currently callers of this method have to filter out `undefined` elements although I added the `notEmpty` filter here. Is there a TS magic to solve this inconvenience?

Reply:
Maybe we should rewrite it so that everybody can understand how it works :-)
FYI https://github.com/Microsoft/TypeScript/issues/12578#issuecomment-263967420